engine/crates/gateway-core/src/lib.rs

```diff
@@ -1,151 +1,127 @@
-use std::{borrow::Cow, collections::HashMap, sync::Arc};
+use std::marker::PhantomData;
+use std::sync::Arc;
 
-use common_types::auth::ExecutionAuth;
 use engine::parser::types::OperationType;
-use futures_util::{future::BoxFuture, FutureExt};
-use http::status::StatusCode;
+use futures_util::FutureExt;
 use runtime::cache::Cache;
+pub use runtime::context::RequestContext;
 use tracing::{info_span, Instrument};
 
 mod admin;
 mod auth;
 mod cache;
 mod executor;
-mod streaming_format;
+mod response;
+pub mod serving;
+mod streaming;
 
-pub use auth::{authorize_request, AuthError};
+pub use auth::{authorize_request, AdminAuthError, AuthError, Authorizer};
 pub use cache::{CacheConfig, CacheControl};
 pub use executor::Executor;
-pub use streaming_format::StreamingFormat;
+pub use response::Response;
+pub use streaming::{encode_stream_response, format::StreamingFormat};
 
-#[async_trait::async_trait]
-pub trait RequestContext: Send + Sync + Sized {
-    fn ray_id(&self) -> &str;
-    // worker requires a 'static future, so there isn't any choice.
-    async fn wait_until(&self, fut: BoxFuture<'static, ()>);
-
-    // Legacy for auth, because it also tries to retrieve it from the query params...
-    fn authorization_header(&self) -> Option<String>;
-    fn header(&self, name: &str) -> Option<String>;
-    fn headers(&self) -> HashMap<String, String>;
+#[derive(thiserror::Error, Debug)]
+pub enum Error {
+    #[error("Bad request: {0}")]
+    BadRequest(String),
+    #[error(transparent)]
+    Cache(runtime::cache::Error),
+    #[error("Serialization Error: {0}")]
+    Serialization(String),
 }
 
-#[async_trait::async_trait]
-pub trait Gateway: Send {
-    // A bit tedious but this allows the caller to have an easier time with its error, response and
-    // context types.
-    type Error: From<Error> + std::error::Error + Send + 'static;
-    type Context: RequestContext;
-    type Response: Response<Context = Self::Context, Error = Self::Error>;
-    type Cache: Cache<Value = engine::Response> + 'static;
-    type Executor: Executor<Context = Self::Context, Error = Self::Error, Response = Self::Response> + 'static;
-
-    fn context(&self) -> &Arc<Self::Context>;
-    fn cache(&self) -> &Arc<Self::Cache>;
-    fn executor(&self) -> &Arc<Self::Executor>;
-    fn cache_config(&self) -> Cow<'_, CacheConfig<'_>>;
-
-    async fn authorize_admin_request(&self, _request: &async_graphql::Request) -> Result<(), AdminAuthError>;
+// A bit tedious but this allows the caller to have an easier time with its error, response and
+// context types.
+pub struct Gateway<Executor, Cache, Context, Response, Error> {
+    executor: Arc<Executor>,
+    cache: Arc<Cache>,
+    cache_config: CacheConfig,
+    authorizer: Box<dyn Authorizer<Context = Context>>,
+    _phantom: PhantomData<fn(Context) -> (Response, Error)>,
```

Comment:
Context appears on that `dyn Authorizor` so you probably don't need it in the phantom

```suggestion
    authorizer: Box<dyn Authorizer<Context = Context>>,
    _phantom: PhantomData<fn() -> (Response, Error)>,
```